src/Clip.cpp

```diff
@@ -1053,7 +1116,250 @@ std::shared_ptr<Frame> Clip::apply_effects(std::shared_ptr<Frame> frame)
+		frame->AddImage(std::shared_ptr<QImage>(source_image));
+	}
+
+	/* ALPHA & OPACITY */
+	if (alpha.GetValue(frame->number) != 1.0)
+	{
+		float alpha_value = alpha.GetValue(frame->number);
+
+		// Get source image's pixels
+		unsigned char *pixels = source_image->bits();
+
+		// Loop through pixels
+		for (int pixel = 0, byte_index=0; pixel < source_image->width() * source_image->height(); pixel++, byte_index+=4)
+		{
+			// Apply alpha to pixel values (since we use a premultiplied value, we must
+			// multiply the alpha with all colors).
+			pixels[byte_index + 0] *= alpha_value;
+			pixels[byte_index + 1] *= alpha_value;
+			pixels[byte_index + 2] *= alpha_value;
+			pixels[byte_index + 3] *= alpha_value;
+		}
+
+		// Debug output
+		ZmqLogger::Instance()->AppendDebugMethod("Clip::apply_keyframes (Set Alpha & Opacity)", "alpha_value", alpha_value, "frame->number", frame->number);
+	}
+
+	/* RESIZE SOURCE IMAGE - based on scale type */
+	QSize source_size = source_image->size();
+	switch (scale)
+	{
+		case (SCALE_FIT): {
+			// keep aspect ratio
+			source_size.scale(width, height, Qt::KeepAspectRatio);
+
+			// Debug output
+			ZmqLogger::Instance()->AppendDebugMethod("Clip::apply_keyframes (Scale: SCALE_FIT)", "frame->number", frame->number, "source_width", source_size.width(), "source_height", source_size.height());
+			break;
+		}
+		case (SCALE_STRETCH): {
+			// ignore aspect ratio
+			source_size.scale(width, height, Qt::IgnoreAspectRatio);
+
+			// Debug output
+			ZmqLogger::Instance()->AppendDebugMethod("Clip::apply_keyframes (Scale: SCALE_STRETCH)", "frame->number", frame->number, "source_width", source_size.width(), "source_height", source_size.height());
+			break;
+		}
+		case (SCALE_CROP): {
+			QSize width_size(width, round(width / (float(source_size.width()) / float(source_size.height()))));
+			QSize height_size(round(height / (float(source_size.height()) / float(source_size.width()))), height);
+
+			// respect aspect ratio
+			if (width_size.width() >= width && width_size.height() >= height)
+				source_size.scale(width_size.width(), width_size.height(), Qt::KeepAspectRatio);
+			else
+				source_size.scale(height_size.width(), height_size.height(), Qt::KeepAspectRatio);
+
+			// Debug output
+			ZmqLogger::Instance()->AppendDebugMethod("Clip::apply_keyframes (Scale: SCALE_CROP)", "frame->number", frame->number, "source_width", source_size.width(), "source_height", source_size.height());
+			break;
+		}
+		case (SCALE_NONE): {
+			// Calculate ratio of source size to project size
+			// Even with no scaling, previews need to be adjusted correctly
+			// (otherwise NONE scaling draws the frame image outside of the preview)
+			float source_width_ratio = source_size.width() / float(width);
+			float source_height_ratio = source_size.height() / float(height);
+			source_size.scale(width * source_width_ratio, height * source_height_ratio, Qt::KeepAspectRatio);
```

Comment:
This is exactly equivalent to:
```cpp
source_size.scale(source_size.width(), source_size.height(), Qt::KeepAspectRatio);
```
The two `width`s and `height`s in the ratios and in the arguments cancel each other out. So, it's a complete no-op.

Reply:
(By canceling, I mean:)

`width` * `source_width_ratio`
= <s>`width`</s> * ( `source_size.width()` / <s>`float(width)`</s> )
= `source_size.width()`